sdk/src/main/java/com/deploygate/sdk/DeployGateClient.java

```diff
@@ -0,0 +1,138 @@
+
+class DeployGateClient {
+    private static final String[] FINGERPRINTS = new String[]{
+            // deploygate release
+            "2f97f647645cb762bf5fc1445599a954e6ad76e7",
+            // mba debug
+            "c1f285f69cc02a397135ed182aa79af53d5d20a1"
+    };
+
+    public final long versionCode;
+    public final boolean isInstalled;
+    private final long featuresFlag;
+
+    @SuppressLint("PackageManagerGetSignatures")
+    DeployGateClient(
+            Context context,
+            String packageName
+    ) {
+        PackageInfo info = null;
+        int flag = PackageManager.GET_META_DATA;
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+            flag |= PackageManager.GET_SIGNING_CERTIFICATES;
+        } else {
+            flag |= PackageManager.GET_SIGNATURES;
+        }
+
+        try {
+            info = context.getPackageManager().getPackageInfo(packageName, flag);
+        } catch (PackageManager.NameNotFoundException e) {
+            Logger.w("deploygate app is not found");
+        }
+
+        if (info != null) {
+            this.isInstalled = checkSignature(info);
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                this.versionCode = info.getLongVersionCode();
+            } else {
+                this.versionCode = info.versionCode;
+            }
+
+            this.featuresFlag = info.applicationInfo.metaData.getLong("com.deploygate.features", 0);
+        } else {
+            this.isInstalled = false;
+            this.versionCode = 0;
+            this.featuresFlag = 0;
+        }
+    }
+
+    boolean isSupported(Compatibility compatibility) {
+        if (featuresFlag <= 0) {
+            switch (compatibility) {
+                case UPDATE_MESSAGE_OF_BUILD:
+                    return versionCode >= Compatibility.ClientVersion.SUPPORT_UPDATE_MESSAGE_OF_BUILD;
+                case SERIALIZED_EXCEPTION:
+                    return versionCode >= Compatibility.ClientVersion.SUPPORT_SERIALIZED_EXCEPTION;
+                default:
+                    return false;
+            }
+        }
+
+        return (featuresFlag & compatibility.bitMask) > 0;
+    }
+
+    /**
+     * Check if the whitelist contains at least one of signing information.
+     * Returning *true* may contain false-positive API 19 or lower. ref: FakeID
+     *
+     * @param info
+     *         com.deploygate's package info
+     *
+     * @return true if at least one of signature is in the whitelist, otherwise false.
+     */
+    private static boolean checkSignature(PackageInfo info) {
+        final MessageDigest md;
+
+        try {
+            md = MessageDigest.getInstance("SHA1");
+        } catch (NoSuchAlgorithmException e) {
+            Logger.e("SHA1 is not supported on this platform?", e);
+            return false;
+        }
+
+        final Signature[] signatures;
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
```

Comment:
`signatures` has been deprecated since Android P. New API signingInfo suggests using getSigningCertificateHistory instead of getApkContentsSigners unless the app has multiple signers. 

BTW, the previous API includes FakeID vulnerability but it had no impact to our app. We haven't enabled any sensitive feature based on the signature match but to reduce the usages of resources like CPU.

Reply:
Wow, I didn't know that vuin. Good catch.